cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
My fix replaces string:iterator by just char pointer (null-terminated c-string), since it is used appropriately.
end can be removed at all, with corresponding changes like
s/charp == end/*charp == 0/

It's successfully compiled and executed (in release), but  debug build is failed with assert on testing stage.